File: 437-path-sum-iii/437-path-sum-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __recurse(self, root, targetSum):
        if root == None:
            return
        
        self.__path.append(root.val)
        
        _sum = 0
        for i in range(len(self.__path) - 1, -1, -1):
            _sum += self.__path[i]
            if _sum == targetSum:
                self.__count += 1
        
        self.__recurse(root.left, targetSum)
        self.__recurse(root.right, targetSum)
        
        self.__path.pop()        
        
        
    def __bruteForce(self, root, targetSum):
        self.__count = 0
        self.__path = list()
        self.__recurse(root, targetSum)
        return self.__count
    
    
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        return self.__bruteForce(root, targetSum)
        

File: 437-path-sum-iii/437-path-sum-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int pathSum(TreeNode* root, int targetSum) {
        return bruteForce(root, targetSum);
    }
private:
    void recurse(TreeNode* root, int targetSum, int& count, vector<int>& path) {
        // base case
        if(root == nullptr)
            return;        
        
        path.push_back(root -> val);
        
        unsigned long int sum = 0;
        for(int i = path.size() - 1; i >= 0; --i) {
            sum += path[i];
            if(sum == targetSum)
                ++count;
        }
        recurse(root -> left, targetSum, count, path);
        recurse(root -> right, targetSum, count, path);
        
        path.pop_back();
    }
    
    
    
    int bruteForce(TreeNode* root, int targetSum) {
        int count = 0;
        vector<int> path;
        recurse(root, targetSum, count, path);
        return count;
    }
};